File: 10-Anamoly Detection/ex10.m
fprintf('Visualizing example dataset for outlier detection.\n\n');
load('ex8data1.mat');

plot(X(:, 1), X(:, 2), 'bx');
axis([0 30 0 30]);
xlabel('Latency (ms)');
ylabel('Throughput (mb/s)');

fprintf('Program paused. Press enter to continue.\n');
pause

function [mu sigma2] = estimateGaussian(X)
  [m, n] = size(X);
  mu = zeros(n, 1);
  sigma2 = zeros(n, 1);
  mu = mean(X);
  sigma2 = var(X).*(m-1)./m;
endfunction

function p = multivariateGaussian(X, mu, Sigma2)
  k = length(mu);
  if (size(Sigma2, 2) == 1) || (size(Sigma2, 1) == 1)
    Sigma2 = diag(Sigma2);
  endif
  X = bsxfun(@minus, X, mu(:)');
  p = (2*pi)^(-k/2)*det(Sigma2)^(-0.5)*exp(-0.5*sum(bsxfun(@times, X*pinv(Sigma2),X), 2));
endfunction

function visualizeFit(X, mu, sigma2)
  [X1,X2] = meshgrid(0:.5:35); 
  Z = multivariateGaussian([X1(:) X2(:)],mu,sigma2);
  Z = reshape(Z,size(X1));
  plot(X(:, 1), X(:, 2),'bx');
  hold on;
  if (sum(isinf(Z)) == 0)
    contour(X1, X2, Z, 10.^(-20:3:0)');
  endif
  hold off;
endfunction

function [bestEpsilon bestF1] = selectThreshold(yval, pval)
  bestEpsilon = 0;
  bestF1 = 0;
  F1 = 0;
  stepsize = (max(pval) - min(pval)) / 1000;
  for epsilon = min(pval):stepsize:max(pval)
    pred = (pval < epsilon);
	  tp = sum((pred == 1) & (yval == 1));
	  fp = sum((pred == 1) & (yval == 0));
	  fn = sum((pred == 0) & (yval == 1));
	  precision = tp/(tp + fp);
	  recall = tp/(tp + fn);
  	F1 = (2*precision*recall)/(precision + recall);
    if F1 > bestF1
       bestF1 = F1;
       bestEpsilon = epsilon;
    endif
  endfor
endfunction
  
fprintf('Visualizing Gaussian fit.\n\n');
[mu sigma2] = estimateGaussian(X);
p = multivariateGaussian(X, mu, sigma2);
visualizeFit(X,  mu, sigma2);
xlabel('Latency (ms)');
ylabel('Throughput (mb/s)');

fprintf('Program paused. Press enter to continue.\n');
pause;

pval = multivariateGaussian(Xval, mu, sigma2);
[epsilon F1] = selectThreshold(yval, pval);
fprintf('Best epsilon found using cross-validation: %e\n', epsilon);
fprintf('Best F1 on Cross Validation Set:  %f\n', F1);

outliers = find(p < epsilon);
hold on
plot(X(outliers, 1), X(outliers, 2), 'ro', 'LineWidth', 2, 'MarkerSize', 10);
hold off

fprintf('Program paused. Press enter to continue.\n');
pause;

load('ex8data2.mat');
[mu sigma2] = estimateGaussian(X);
p = multivariateGaussian(X, mu, sigma2);
pval = multivariateGaussian(Xval, mu, sigma2);
[epsilon F1] = selectThreshold(yval, pval);

fprintf('Best epsilon found using cross-validation: %e\n', epsilon);
fprintf('Best F1 on Cross Validation Set:  %f\n', F1);
fprintf('# Outliers found: %d\n', sum(p < epsilon));
pause

File: 10-Anamoly Detection/ex10.py
import numpy as np
import scipy.io as io
import matplotlib.pyplot as plt

def plot_data(X):
    plt.plot(X[:, 0], X[:, 1], 'bx')
    plt.xlim(0, 30)
    plt.ylim(0, 30)
    plt.xlabel('Latency (ms)')
    plt.ylabel('Throughput (mb/s)')

def visualize_fit(X, mu, sigma2):
    l = np.arange(0, 30.5, 0.5)
    X1, X2 = np.meshgrid(l, l)
    X_tmp = np.vstack((X1.ravel(), X2.ravel())).T
    Z = multivariate_gaussian(X_tmp, mu, sigma2)
    Z.resize(X1.shape)
    plt.plot(X[:, 0], X[:, 1], 'bx')
    if np.sum(np.isinf(Z)) == 0:
        plt.contour(X1, X2, Z, 10.0 ** np.arange(-20, 0, 3))

def estimate_gaussian(X):
    mu = np.mean(X, axis=0)
    sigma2 = np.var(X, axis=0)
    return mu, sigma2

def multivariate_gaussian(X, mu, sigma2):
    m,n = np.shape(X)
    if len(sigma2.shape) == 1:
        sigma2 = np.diag(sigma2)
    p = np.power((2*np.pi),(-n/2))*np.power(np.linalg.det(sigma2),-0.5)*\
        np.exp(-0.5*np.sum(((X-mu).dot(np.linalg.pinv(sigma2))* (X-mu)),axis=1))
    return p

def select_threshold(y_val, p_val):
    step_size = (np.max(p_val) - np.min(p_val)) / 1000
    best_epsilon = 0.0
    best_F1 = 0.0
    for epsilon in np.arange(min(p_val), max(p_val), step_size):
        predictions = p_val < epsilon
        tp = np.sum(predictions[np.nonzero(y_val == True)])
        fp = np.sum(predictions[np.nonzero(y_val == False)])
        fn = np.sum(y_val[np.nonzero(predictions == False)] == True)
        if tp != 0:
            prec = 1.0 * tp / (tp + fp)
            rec = 1.0 * tp / (tp + fn)
            F1 = 2.0 * prec * rec / (prec + rec)
            if F1 > best_F1:
                best_F1 = F1
                best_epsilon = epsilon
    return best_epsilon, best_F1

if __name__ == "__main__":
    # ================== Part 1: Load Example Dataset  ===================
    print ('Visualizing example dataset for outlier detection.')
    data = io.loadmat('ex8data1.mat')
    X = data['X']
    X_val = data['Xval']
    y_val = data['yval'].ravel()
    plt.figure(1)
    plot_data(X)

    # ================== Part 2: Estimate the dataset statistics ===================
    print ('Visualizing Gaussian fit.')
    mu, sigma2 = estimate_gaussian(X)
    p = multivariate_gaussian(X, mu, sigma2)
    plt.figure(2)
    visualize_fit(X, mu, sigma2)

    # ================== Part 3: Find Outliers ===================
    p_val = multivariate_gaussian(X_val, mu, sigma2)
    epsilon, F1 = select_threshold(y_val, p_val)

    print ('Best epsilon found using cross-validation:', epsilon)
    print ('Best F1 on Cross Validation Set:', F1)

    outliers = np.nonzero(p < epsilon)
    plt.figure(2)
    plt.scatter(X[outliers, 0], X[outliers, 1], facecolors='none', edgecolors='r', s=100)
    plt.show()

    # ================== Part 4: Multidimensional Outliers ===================
    # Loads the second dataset.
    mat_data = io.loadmat('ex8data2.mat')
    X = mat_data['X']
    X_val = mat_data['Xval']
    y_val = mat_data['yval'].ravel()

    mu, sigma2 = estimate_gaussian(X)
    p = multivariate_gaussian(X, mu, sigma2)
    p_val = multivariate_gaussian(X_val, mu, sigma2)
    epsilon, F1 = select_threshold(y_val, p_val)
    print('Best epsilon found using cross-validation:', epsilon)
    print('Best F1 on Cross Validation Set:', F1)
    print ('# Outliers found:', np.sum(p < epsilon))